File: src/fixtures/MultiDimMKII.js
/*
 * Copyright (c) 2023, Bastiaan van der Plaat <bastiaan.v.d.plaat@gmail.com>
 *
 * SPDX-License-Identifier: MIT
 */

import Fixture from './Fixture.js';

export default class MultiDimMKII extends Fixture {
    constructor(name, addr, options = {}) {
        super();
        this.type = 'multidim';
        this.name = name;
        this.addr = addr;
        this.state = new Array(4).fill(false);
        if ('labels' in options) {
            this.labels = options.labels;
        }
    }

    update(dmx) {
        dmx[this.addr + 0] = this.state[0] ? 255 : 0;
        dmx[this.addr + 1] = this.state[1] ? 255 : 0;
        dmx[this.addr + 2] = this.state[2] ? 255 : 0;
        dmx[this.addr + 3] = this.state[3] ? 255 : 0;
    }
}

File: bassielight/src/fixtures/multidim_mkii.rs
/*
 * Copyright (c) 2023-2024, Bastiaan van der Plaat <bastiaan.v.d.plaat@gmail.com>
 *
 * SPDX-License-Identifier: MIT
 */

use crate::devices::DMXEmitter;
use crate::fixtures::{Fixture, SwitchFixture};
use std::any::Any;

pub struct MultiDimMKII {
    addr: usize,

    state: [bool; 4],
}

impl MultiDimMKII {
    pub fn new(addr: usize) -> Self {
        Self {
            addr: addr,
            state: [false; 4],
        }
    }
}

impl Fixture for MultiDimMKII {
    fn as_any(&mut self) -> &mut dyn Any {
        self
    }

    fn tick(&mut self, dmx: &mut dyn DMXEmitter) {
        for i in 0..self.state.len() {
            dmx.write(self.addr + i, if self.state[i] { 255 } else { 0 });
        }
    }
}

impl SwitchFixture for MultiDimMKII {
    fn state(&self, index: usize) -> bool {
        self.state[index]
    }
    fn set_state(&mut self, index: usize, state: bool) {
        self.state[index] = state;
    }
    fn toggle(&mut self, index: usize) {
        self.state[index] = !self.state[index];
    }
}
